tools/codegen/dest/ufunc.py

```diff
@@ -0,0 +1,440 @@
+from dataclasses import dataclass, field
+from typing import Union, Optional, List, Tuple, Dict, Sequence
+from tools.codegen.api.translate import translate
+from tools.codegen.utils import Target
+from tools.codegen.model import NativeFunctionsGroup, ScalarType, UfuncKey, DispatchKey, BaseType, BaseTy, Argument
+import tools.codegen.api.ufunc as ufunc
+from tools.codegen.api.ufunc import UfunctorBindings
+import tools.codegen.api.structured as structured
+from tools.codegen.api.types import StructuredImplSignature, scalar_t, opmath_t, Binding, CType, BaseCType, Expr, NamedCType, ScalarTypeToCppMapping, scalarT, VectorizedCType
+from tools.codegen.context import with_native_function
+from copy import copy
+
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+#
+#                                  CUDA STUFF
+#
+# ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~ #
+
+# NB: not bothering to generate dispatch stub forward declaration in header,
+# we can just paste it whereever necessary
+
+# TODO: use BackendIndex
+# dispatch_key: DispatchKey  # only CPU/CUDA right now
+
+
+# Represents functors for implementing CUDA ufuncs.
+# Functors are templated by scalar_t because when USERS instantiate functors
+# they are templated.  A functor looks something like this:
+#
+#
+@dataclass(frozen=True)
+class UfunctorSignature:
+    g: NativeFunctionsGroup
+    scalar_tensor: Optional[int]
```

Comment:
nit: maybe `scalar_tensor_idx`? I was confused reading this until I saw where it was used